File: backend/core-api/src/main/java/io/basearchitecture/core/config/AiServiceClientProperties.java
package io.basearchitecture.core.config;

import java.net.URI;
import java.time.Duration;
import java.util.Objects;

import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties(prefix = "app.ai-service")
public record AiServiceClientProperties(
        URI baseUrl,
        Duration connectTimeout,
        Duration readTimeout
) {

    public AiServiceClientProperties {
        Objects.requireNonNull(baseUrl, "app.ai-service.base-url must be configured");
        Objects.requireNonNull(connectTimeout, "app.ai-service.connect-timeout must be configured");
        Objects.requireNonNull(readTimeout, "app.ai-service.read-timeout must be configured");

        validateBaseUrl(baseUrl);
        validatePositive(connectTimeout, "app.ai-service.connect-timeout");
        validatePositive(readTimeout, "app.ai-service.read-timeout");
    }

    private static void validateBaseUrl(URI baseUrl) {
        String scheme = baseUrl.getScheme();
        boolean supportedScheme = "http".equalsIgnoreCase(scheme) || "https".equalsIgnoreCase(scheme);
        if (!baseUrl.isAbsolute() || !supportedScheme || baseUrl.getHost() == null) {
            throw new IllegalArgumentException(
                    "app.ai-service.base-url must be an absolute HTTP(S) URI with a host");
        }

        int port = baseUrl.getPort();
        if (port != -1 && (port < 1 || port > 65_535)) {
            throw new IllegalArgumentException(
                    "app.ai-service.base-url port must be between 1 and 65535");
        }

        String path = baseUrl.getPath();
        if ((path != null && !path.isEmpty() && !"/".equals(path))
                || baseUrl.getQuery() != null
                || baseUrl.getFragment() != null
                || baseUrl.getUserInfo() != null) {
            throw new IllegalArgumentException(
                    "app.ai-service.base-url must contain only scheme, host, and optional port");
        }
    }

    private static void validatePositive(Duration duration, String propertyName) {
        if (duration.isZero() || duration.isNegative()) {
            throw new IllegalArgumentException(propertyName + " must be greater than zero");
        }
    }
}

File: backend/core-api/src/main/kotlin/ai/govbiz/core/config/AiServiceClientProperties.kt
package ai.govbiz.core.config

import java.net.URI
import java.time.Duration
import org.springframework.boot.context.properties.ConfigurationProperties

@ConfigurationProperties(prefix = "app.ai-service")
data class AiServiceClientProperties(
    val baseUrl: URI,
    val connectTimeout: Duration,
    val readTimeout: Duration,
) {

    init {
        validateBaseUrl(baseUrl)
        validatePositive(connectTimeout, "app.ai-service.connect-timeout")
        validatePositive(readTimeout, "app.ai-service.read-timeout")
    }

    companion object {
        private fun validateBaseUrl(baseUrl: URI) {
            val scheme = baseUrl.scheme
            val supportedScheme =
                "http".equals(scheme, ignoreCase = true) ||
                    "https".equals(scheme, ignoreCase = true)
            if (!baseUrl.isAbsolute || !supportedScheme || baseUrl.host == null) {
                throw IllegalArgumentException(
                    "app.ai-service.base-url must be an absolute HTTP(S) URI with a host",
                )
            }

            val port = baseUrl.port
            if (port != -1 && (port < 1 || port > 65_535)) {
                throw IllegalArgumentException(
                    "app.ai-service.base-url port must be between 1 and 65535",
                )
            }

            val path = baseUrl.path
            if (
                (path != null && path.isNotEmpty() && path != "/") ||
                baseUrl.query != null ||
                baseUrl.fragment != null ||
                baseUrl.userInfo != null
            ) {
                throw IllegalArgumentException(
                    "app.ai-service.base-url must contain only scheme, host, and optional port",
                )
            }
        }

        private fun validatePositive(duration: Duration, propertyName: String) {
            if (duration.isZero || duration.isNegative) {
                throw IllegalArgumentException("$propertyName must be greater than zero")
            }
        }
    }
}
